### PR TITLE
Added small set of clang-tidy modernize-* checks

### PR DESCRIPTION
**Pull request by gnawme**
Thanks a lot for making a contribution to SFML! 🙂

Before you create the pull request, we ask you to check the follow boxes. (For small changes not everything needs to ticked, but the more the better!)

* [x] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
* [x] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [ ] Have you provided some example/test code for your changes?
* [ ] If you have additional steps which need to be performed list them as tasks!

----

## Description

Originally opened to address #2312, but #2321 is addressing it.

However, this also adds a small subset of `clang-tidy` `modernize-*` checks to the `.clang-tidy`:
`  modernize-use-equals-default,
  modernize-use-equals-delete,
  modernize-use-nullptr,
  modernize-use-override,
  modernize-use-using,
`

This PR is related to the issue #2312 

## Tasks

* [x] Tested on Linux
* [ ] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

Describe how to best test these changes. Please provide a [minimal, complete and verifiable example](https://stackoverflow.com/help/mcve) if possible, you can use the follow template as a start:

```
#include <SFML/Graphics.hpp>

int main()
{
    sf::Vector2u modeSize{200, 200};
    sf::RenderWindow window(sf::VideoMode(modeSize), "SFML works!");
    sf::CircleShape shape(100.f);
    shape.setFillColor(sf::Color::Green);

    while (window.isOpen())
    {
        sf::Event event;
        while (window.pollEvent(event))
        {
            if (event.type == sf::Event::Closed)
                window.close();
        }

        window.clear();
        window.draw(shape);
        window.display();
    }

    return 0;
}

g++ main.cpp -o sfml-app -I ./include/ -L./build/lib -lsfml-audio -lsfml-graphics -lsfml-system -lsfml-window

```

### CODE REVIEW

**Comment by ChrisThrasher** [comment]
I'm glad someone's adding more checks! Were all of these changes motivated by clang-tidy errors you got after adding those checks?

It looks like every single CI build job is failing. Please be sure to build this code on your own machine first to make sure it works on at least one OS before pushing.

The `nullptr` stuff is great and we could easily merge that if it was in its own PR. The other stuff dealing with destructors and other special members is a bit touchier since [this comment](https://github.com/SFML/SFML/issues/2312#issuecomment-1366776905) pointed out something I didn't know when writing issue #2312.

**Comment by gnawme** [comment]
> I'm glad someone's adding more checks! Were all of these changes motivated by clang-tidy errors you got after adding those checks?
> 
> It looks like every single CI build job is failing. Please be sure to build this code on your own machine first to make sure it works on at least one OS before pushing.
> 
> The `nullptr` stuff is great and we could easily merge that if it was in its own PR. The other stuff dealing with destructors and other special members is a bit touchier since [this comment](https://github.com/SFML/SFML/issues/2312#issuecomment-1366776905) pointed out something I didn't know when writing issue #2312.

Yes, the initial commit addressed the `clang-tidy` errors, the last commit was to address #2312.

I built successfully on my machine (Ubuntu 20), and even ran `clang-format` on the files I touched, so I'll look at the log and see what the complaint is.

**Comment by ChrisThrasher** [comment]
Please build the `format` target which is the preferred way to run clang-format on SFML. I see the formatting CI jobs are failing.

**Comment by gnawme** [comment]
> Please build the `format` target which is the preferred way to run clang-format on SFML. I see the formatting CI jobs are failing.

Thanks, no complaints when building the `format` target with my current push.

File: src/SFML/Graphics/ImageLoader.hpp [hunk 1]
```diff
@@ -121,25 +139,6 @@ class ImageLoader
...
-    /// \brief Deleted copy constructor
-    ///
-    ////////////////////////////////////////////////////////////
-    ImageLoader(const ImageLoader&) = delete;
-
-    ////////////////////////////////////////////////////////////
-    /// \brief Deleted copy assignment
-    ///
-    ////////////////////////////////////////////////////////////
-    ImageLoader& operator=(const ImageLoader&) = delete;
```
Comment (ChrisThrasher) [inline src/SFML/Graphics/ImageLoader.hpp:142]: Did clang-tidy instruct you to move these to the header file?

Comment (gnawme) [reply src/SFML/Graphics/ImageLoader.hpp:142]: '
[1m/home/norme/Projects/SFML/src/SFML/Graphics/ImageLoader.hpp:136:5: [0m[0;1;31merror: [0m[1mdeleted member function should be public [modernize-use-equals-delete,-warnings-as-errors][0m
    ImageLoader(const ImageLoader&) = delete;
[0;1;32m    ^
[0m[1m/home/norme/Projects/SFML/src/SFML/Graphics/ImageLoader.hpp:142:18: [0m[0;1;31merror: [0m[1mdeleted member function should be public [modernize-use-equals-delete,-warnings-as-errors][0m
    ImageLoader& operator=(const ImageLoader&) = delete;
[0;1;32m                 ^
'

Comment (ChrisThrasher) [reply src/SFML/Graphics/ImageLoader.hpp:142]: I was confusing this with other files where code did move from .cpp files to .hpp files. In some cases the definition of those special member functions was kept in the .cpp for a reason so I'd prefer we not change that in this PR. The changes in sf::String demonstrate what I'm talking about.

Comment (gnawme) [reply src/SFML/Graphics/ImageLoader.hpp:142]: > I was confusing this with other files where code did move from .cpp files to .hpp files. In some cases the definition of those special member functions was kept in the .cpp for a reason so I'd prefer we not change that in this PR. The changes in sf::String demonstrate what I'm talking about.

Declaring `String()` in the header, then defining `String() = default` in the source seems redundant; is that what you would prefer?

**Comment by ChrisThrasher** [comment]
There's a bug in our clang-tidy CI job that I need to fix that is allowing the jobs to pass even though there are errors. If you read the logs you'll see there are lots of unaddressed clang-tidy errors in there. I'll fix this and then we can rebase this PR once that's merged and see what all is failing more clearly.

**Comment by ChrisThrasher** [comment]
See https://github.com/SFML/SFML/pull/2337

File: src/SFML/Graphics/ImageLoader.hpp (hunk 1 above)
Comment (ChrisThrasher) [reply src/SFML/Graphics/ImageLoader.hpp:142]: Yes it seems odd and I'm not sure it's my personal ideal but for now I'd rather not go moving those definitions around so we can keep the scope of this PR manageable.

**Comment by gnawme** [comment]
> There's a bug in our clang-tidy CI job that I need to fix that is allowing the jobs to pass even though there are errors. If you read the logs you'll see there are lots of unaddressed clang-tidy errors in there. I'll fix this and then we can rebase this PR once that's merged and see what all is failing more clearly.

OK, this PR is small enough to abandon and just do the `modernize-` checks on top of your changes.

**Comment by ChrisThrasher** [comment]
Alright that fixed has been merged. Pull and rebase on master to get those changes.

**Comment by gnawme** [comment]
Superseded by #2339

**Comment by ChrisThrasher** [comment]
You didn't need to make a new PR. You could have just amended this one.

**Comment by gnawme** [comment]
> You didn't need to make a new PR. You could have just amended this one.

Sorry, I moved a few `=delete` methods from source to header, I thought this would be easier than trying to hunt them all down

**Pull request closed without merging**